src/documents/models.py

```diff
@@ -314,6 +391,93 @@ def thumbnail_file(self):
     def thumbnail_url(self):
         return reverse("fetch", kwargs={"kind": "thumb", "pk": self.pk})
 
+    def set_filename(self, filename):
+        if os.path.isfile(Document.filename_to_path(filename)):
+            self.filename = filename
+
+
+def try_delete_empty_directories(directory):
+    # Go up in the directory hierarchy and try to delete all directories
+    directory = os.path.normpath(directory)
+    root = os.path.normpath(Document.filename_to_path(""))
+
+    while directory != root:
+        # Try to delete the current directory
+        try:
+            os.rmdir(directory)
+        except os.error:
+            # Directory not empty, no need to go further up
+            return
+
+        # Cut off actual directory and go one level up
+        directory, _ = os.path.split(directory)
+        directory = os.path.normpath(directory)
```

Comment:
I think this code would really profit from using pathlib. E.g. we could iterate over  `https://docs.python.org/3/library/pathlib.html?highlight=pathlib#pathlib.PurePath.parents` instead of using `while` here.

But we don't use pathlib everywhere else, so probably not the right thing for this PR.

Reply:
I am happy to migrate to path lib, but it is not really used anywhere in paperless. What do you think about doing this change?

Reply:
While I agree that pathlib seems to be _the_ way to go when handling paths, I don't think we have to block this PR on that. Maybe it makes sense to refactor this separately?

Reply:
Yes, we should totally do it in probably multiple different PRs, so it's easier to work through.